<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.1 plus MathML 2.0//EN" "http://www.w3.org/Math/DTD/mathml2/xhtml-math11-f.dtd"><html xmlns="http://www.w3.org/1999/xhtml"><!--This file was converted to xhtml by LibreOffice - see http://cgit.freedesktop.org/libreoffice/core/tree/filter/source/xslt for the code.--><head profile="http://dublincore.org/documents/dcmi-terms/"><meta http-equiv="Content-Type" content="application/xhtml+xml; charset=utf-8"/><title xml:lang="en-US">- no title specified</title><meta name="DCTERMS.title" content="" xml:lang="en-US"/><meta name="DCTERMS.language" content="en-US" scheme="DCTERMS.RFC4646"/><meta name="DCTERMS.source" content="http://xml.openoffice.org/odf2xhtml"/><meta name="DCTERMS.issued" content="2019-12-13T10:44:36.106627569" scheme="DCTERMS.W3CDTF"/><meta name="DCTERMS.modified" content="2019-12-13T11:49:57.747532749" scheme="DCTERMS.W3CDTF"/><meta name="DCTERMS.provenance" content="" xml:lang="en-US"/><meta name="DCTERMS.subject" content="," xml:lang="en-US"/><link rel="schema.DC" href="http://purl.org/dc/elements/1.1/" hreflang="en"/><link rel="schema.DCTERMS" href="http://purl.org/dc/terms/" hreflang="en"/><link rel="schema.DCTYPE" href="http://purl.org/dc/dcmitype/" hreflang="en"/><link rel="schema.DCAM" href="http://purl.org/dc/dcam/" hreflang="en"/><style type="text/css">
    @page {  }
    table { border-collapse:collapse; border-spacing:0; empty-cells:show }
    td, th { vertical-align:top; font-size:12pt;}
    h1, h2, h3, h4, h5, h6 { clear:both;}
    ol, ul { margin:0; padding:0;}
    li { list-style: none; margin:0; padding:0;}
    /* "li span.odfLiEnd" - IE 7 issue*/
    li span. { clear: both; line-height:0; width:0; height:0; margin:0; padding:0; }
    span.footnodeNumber { padding-right:1em; }
    span.annotation_style_by_filter { font-size:95%; font-family:Arial; background-color:#fff000;  margin:0; border:0; padding:0;  }
    span.heading_numbering { margin-right: 0.8rem; }* { margin:0;}
    .P1 { font-size:13pt; line-height:115%; margin-bottom:0.247cm; margin-top:0cm; font-family:Liberation Serif; writing-mode:page; }
    .P10 { font-size:12pt; line-height:150%; margin-bottom:0cm; margin-top:0cm; font-family:Liberation Serif; writing-mode:page; }
    .P11 { font-size:12pt; line-height:150%; margin-bottom:0cm; margin-top:0cm; font-family:Liberation Serif; writing-mode:page; }
    .P12 { font-size:13pt; line-height:150%; margin-bottom:0cm; margin-top:0cm; font-family:Liberation Serif; writing-mode:page; }
    .P13 { font-size:13pt; line-height:150%; margin-bottom:0cm; margin-top:0cm; font-family:Liberation Serif; writing-mode:page; font-style:italic; }
    .P14 { font-size:13pt; line-height:150%; margin-bottom:0.649cm; margin-top:0.402cm; font-family:Liberation Serif; writing-mode:page; font-weight:bold; }
    .P15 { font-size:115%; font-weight:bold; margin-bottom:0.212cm; margin-top:0.353cm; font-family:Liberation Sans; writing-mode:page; text-align:center ! important; }
    .P16 { font-size:101%; font-weight:bold; margin-bottom:0.212cm; margin-top:0.247cm; font-family:Liberation Sans; writing-mode:page; }
    .P17 { font-size:101%; font-weight:bold; margin-bottom:0.212cm; margin-top:0.247cm; font-family:Liberation Sans; writing-mode:page; }
    .P18 { font-size:101%; font-weight:bold; margin-bottom:0.111cm; margin-top:0.146cm; font-family:Liberation Sans; writing-mode:page; line-height:150%; }
    .P2 { font-size:13pt; line-height:115%; margin-bottom:0.247cm; margin-top:0cm; font-family:Liberation Serif; writing-mode:page; }
    .P3 { font-size:12pt; line-height:115%; margin-bottom:0.247cm; margin-top:0cm; font-family:Liberation Serif; writing-mode:page; }
    .P4 { font-size:12pt; line-height:150%; margin-bottom:0.448cm; margin-top:0.201cm; font-family:Liberation Serif; writing-mode:page; }
    .P5 { font-size:12pt; line-height:150%; margin-bottom:0.448cm; margin-top:0.201cm; font-family:Liberation Serif; writing-mode:page; }
    .P6 { font-size:12pt; line-height:150%; margin-bottom:0.448cm; margin-top:0.201cm; font-family:Liberation Serif; writing-mode:page; }
    .P7 { font-size:13pt; line-height:150%; margin-bottom:0.448cm; margin-top:0.201cm; font-family:Liberation Serif; writing-mode:page; }
    .P8 { font-size:13pt; line-height:150%; margin-bottom:0.046cm; margin-top:0cm; font-family:Liberation Serif; writing-mode:page; }
    .P9 { font-size:13pt; line-height:150%; margin-bottom:0.046cm; margin-top:0cm; font-family:Liberation Serif; writing-mode:page; font-style:italic; }
    .Bullet_20_Symbols { font-family:OpenSymbol; }
    .T10 { font-size:13pt; font-weight:normal; }
    .T11 { font-size:13pt; font-weight:normal; }
    .T12 { font-size:13pt; font-weight:normal; }
    .T13 { font-size:13pt; }
    .T14 { font-size:13pt; }
    .T19 { font-weight:normal; }
    .T22 { font-weight:normal; }
    .T3 { font-size:13pt; }
    .T4 { font-size:13pt; }
    .T5 { font-size:13pt; }
    .T6 { font-size:13pt; font-weight:bold; }
    .T7 { font-size:13pt; font-weight:bold; }
    .T8 { font-size:13pt; font-weight:bold; }
    .T9 { font-size:13pt; font-weight:normal; }
    /* ODF styles with no properties representable as CSS */
    .T1 .T2 .T23  { }
    </style></head><body dir="ltr" style="max-width:21.001cm;margin-top:2cm; margin-bottom:2cm; margin-left:2cm; margin-right:2cm; "><h2 class="P15"><a id="a__Project_n°1___Genetic_of_Populations__Wright-Fisher__ReadMe"><span/></a>        <span class="T1">Project n°1 : Genetic of Populations (Wright-Fisher) ReadMe</span></h2><p class="P3"><span class="T4">[Team n°1]</span></p><p class="P3"><span class="T4"/></p><p class="P3"><span class="T4"/></p><h3 class="P17"><a id="a__I__What_the_program_does"><span/></a>I. What the program does</h3><p class="P5"><span class="T13">        The main goal of this program is to simulate the temporal evolution of a population represented by N individuals with a genome. In order to do that, we will follow the model of Wright-Fisher. There is a fixed number A of alleles in the population and each individual has a type of allele. At each generation, we will sort a new population with a size N among the individuals of the last generation according to the multinomial distribution. The alleles of this population can mutate during the simulation (with mutation coefficients) or can be more or less selec</span><span class="T14">ted (with fitness coefficients).</span></p><h3 class="P18"><a id="a__II__The_software_you_need_to_install"><span/></a><span class="T2">II. The software you need to install</span></h3><ul><li><p class="P2" style="margin-left:0cm;"><span class="Bullet_20_Symbols" style="display:block;float:left;min-width:0,635cm;">•</span>C<span class="T2">M</span>ake (compilation)<span class="odfLiEnd"/> </p></li><li><p class="P2" style="margin-left:0cm;"><span class="Bullet_20_Symbols" style="display:block;float:left;min-width:0,635cm;">•</span>Doxygen (documentation)<span class="odfLiEnd"/> </p></li><li><p class="P2" style="margin-left:0cm;"><span class="Bullet_20_Symbols" style="display:block;float:left;min-width:0,635cm;">•</span>G<span class="T2">t</span>est (tests)<span class="odfLiEnd"/> </p></li><li><p class="P2" style="margin-left:0cm;"><span class="Bullet_20_Symbols" style="display:block;float:left;min-width:0,635cm;">•</span>TCLAP (command on terminal)<span class="odfLiEnd"/> </p></li></ul><p class="P1"> </p><h3 class="P16"><a id="a__III__How_to_execute_the_program"><span/></a>II<span class="T23">I</span>. How to execute the program</h3><p class="P4"><span> <span class="T3">You have two possible ways to perform the simulation : with our without a fasta file.      If you do not want to work with a fasta file you need to </span><span class="T5">enter all the factors the simulation needs. : </span></span></p><ul><li><p class="P10" style="margin-left:0cm;"><span class="Bullet_20_Symbols" style="display:block;float:left;min-width:0,635cm;">•</span><span class="T5">The time of simulation or the number of generations with </span><span class="T6">-T</span><span class="T9"> (</span><span class="T10">default value</span><span class="T9"> : 10)                          </span><span class="odfLiEnd"/> </p></li><li><p class="P10" style="margin-left:0cm;"><span class="Bullet_20_Symbols" style="display:block;float:left;min-width:0,635cm;">•</span><span class="T9">The number of repetitions of the simulation with </span><span class="T6">-R</span><span class="T9"> (</span><span class="T10">default </span><span class="T9">value : 2)                                      </span><span class="odfLiEnd"/> </p></li><li><p class="P10" style="margin-left:0cm;"><span class="Bullet_20_Symbols" style="display:block;float:left;min-width:0,635cm;">•</span><span class="T9">The size of the population with </span><span class="T6">-N</span><span class="T9"> (default value : </span><span class="T10">100)                                                                      </span><span class="odfLiEnd"/> </p></li><li><p class="P10" style="margin-left:0cm;"><span class="Bullet_20_Symbols" style="display:block;float:left;min-width:0,635cm;">•</span><span class="T10">The number of alleles of the population with </span><span class="T7">-A </span><span class="T10">(default value : 2)                                                  </span><span class="odfLiEnd"/> </p></li><li><p class="P10" style="margin-left:0cm;"><span class="Bullet_20_Symbols" style="display:block;float:left;min-width:0,635cm;">•</span><span class="T10">The frequencies of the alleles with </span><span class="T7">-f </span><span class="T10">(no default value) -</span><span class="odfLiEnd"/> </p></li><li><p class="P10" style="margin-left:0cm;"><span class="Bullet_20_Symbols" style="display:block;float:left;min-width:0,635cm;">•</span><span class="T10">The fitness coefficient of the alleles with </span><span class="T7">-S</span><span class="T10"> (default value : </span><span class="T12">0 for each allele</span><span class="T10">)</span><span class="odfLiEnd"/> </p></li></ul><p class="P6"/><p class="P7">If you to use a fasta file to perform the simulation, you need to enter these factors : </p><ul><li><p class="P11" style="margin-left:0cm;"><span class="Bullet_20_Symbols" style="display:block;float:left;min-width:0,635cm;">•</span><span class="T5">The time of simulation or the number of generations with </span><span class="T6">-T</span><span class="T9"> (</span><span class="T10">default value</span><span class="T9"> : 10)                           </span><span class="odfLiEnd"/> </p></li><li><p class="P11" style="margin-left:0cm;"><span class="Bullet_20_Symbols" style="display:block;float:left;min-width:0,635cm;">•</span><span class="T9">The number of repetitions of the simulation with </span><span class="T6">-R</span><span class="T9"> (</span><span class="T10">default </span><span class="T9">value : 2)                                           </span><span class="odfLiEnd"/> </p></li><li><p class="P11" style="margin-left:0cm;"><span class="Bullet_20_Symbols" style="display:block;float:left;min-width:0,635cm;">•</span><span class="T11">The name of the fasta file with </span><span class="T8">-F</span><span class="T11"> (default value : ../tests/test_for_retrieveData.fasta)                      </span><span class="odfLiEnd"/> </p></li><li><p class="P11" style="margin-left:0cm;"><span class="Bullet_20_Symbols" style="display:block;float:left;min-width:0,635cm;">•</span><span class="T11">The marks of the 3 nucleotides of the alleles with </span><span class="T8">-m</span><span class="T11"> (no default value)                                            </span><span class="odfLiEnd"/> </p></li><li><p class="P11" style="margin-left:0cm;"><span class="Bullet_20_Symbols" style="display:block;float:left;min-width:0,635cm;">•</span><span class="T11">The mutation rate of the alleles with </span><span class="T8">-M</span><span class="T11"> (</span><span class="T10">default value : </span><span class="T12">0 for each allele</span><span class="T11">)                                                               </span><span class="odfLiEnd"/> </p></li><li><p class="P11" style="margin-left:0cm;"><span class="Bullet_20_Symbols" style="display:block;float:left;min-width:0,635cm;">•</span><span class="T11">The mutations sites with </span><span class="T8">-m</span><span class="T11"> (</span><span class="T10">default value : </span><span class="T12">0 for each allele</span><span class="T11">)                                                                      </span><span class="odfLiEnd"/> </p></li><li><p class="P11" style="margin-left:0cm;"><span class="Bullet_20_Symbols" style="display:block;float:left;min-width:0,635cm;">•</span><span class="T11">The </span><span class="T10">fitness coefficient of the alleles with </span><span class="T7">-S</span><span class="T10"> (default value : </span><span class="T12">0 for each allele</span><span class="T10">)                                                          </span><span class="odfLiEnd"/> </p></li><li><p class="P11" style="margin-left:0cm;"><span class="Bullet_20_Symbols" style="display:block;float:left;min-width:0,635cm;">•</span><span class="T11">The delta coefficient of the mutation with </span><span class="T8">-d</span><span class="T11"> (default value : 1/3)                                                     </span><span class="odfLiEnd"/> </p></li></ul><p class="P14">Examples of command : </p><p class="P13"><span class="T22">-  Without a fasta file and with selection :</span></p><p class="P12"><span class="T22">./PopulationGenetic -R 4 -T 100 -A 4 -f 0.1 -f 0.1 -f 0.3 -f 0.5 -N 100 -S -1 -S 0 -S -1 -S 0.4</span></p><p class="P12"><span class="T22"/></p><p class="P9"><span class="T22">-  With a fasta file and mutations and without selection :</span></p><p class="P8"><span class="T22">./PopulationGenetic -F ../test.fa -m 2 -m 4 -m 8 -T 100 -R 2 -M 0.01 -M 0.02 -M 0.5 -m 2 -m 4 -m 8 -d 0.5                                                                                            </span></p><p class="P7"><span class="T19">               </span></p></body></html>